Added cassandra plugin to access metrics using jolokia and push them …

subhachandrachandra opened the pull request:
This plugin is modeled after the jolokia plugin but has custom parsing logic for the json reply that is specific to cassandra. The configuration is bit simpler as it extracts metric names from the json reply. It also uses multiple metric names like "cassandraClient", "cassandraTable", "javaGarbageCollector" like the redis plugin instead of adding everything under one metric.

It has been tested against Cassandra3.x.

subhachandrachandra commented:
Not sure what I need to do here. CircleCI failed in a kafka test.

ok  	github.com/influxdata/telegraf/plugins/outputs/graphite	1.144s
ok  	github.com/influxdata/telegraf/plugins/outputs/influxdb	1.029s
--- FAIL: TestConnectAndWrite (0.80s)
	<autogenerated>:31:                         
	Error Trace:	kafka_test.go:30		
	Error:		Received unexpected error "FAILED to send kafka message: kafka server: In the middle of a leadership election, there is currently no leader for this partition and hence it is unavailable for writes.\n"
FAIL
FAIL	github.com/influxdata/telegraf/plugins/outputs/kafka	0.823s

sparrc commented on plugins/inputs/cassandra/README.md:
````diff
@@ -0,0 +1,41 @@
+Given a configuration like:
+
+```ini
+[cassandra]
````
this is not a valid telegraf configuration anymore, checkout a more recent version of the config file, it should look more like this:

```
[[inputs.cassandra]]
  ...
```

sparrc commented:
servers & metrics should just be regular lists, so that configuration will look like this:

```
[[inputs.cassandra]]
  servers = [...]
  metrics = ["/heap..." ... ]
```

subhachandrachandra commented:
The metrics format has been bothering me for a while. Will fix that. "servers" will be a bit tricky to add user names and passwords. Will have to figure out a format for it.

sparrc commented:
in that case servers may need to be as it is, that's fine

sparrc commented:
@subhachandrachandra why not use the jolokia plugin directly?

subhachandrachandra commented:
@sparcc I did start with the jolokia plugin as described at https://github.com/chrusty/telegraf-cassandra.
I hit the same problems as mentioned by @chrusty. The jolokia plugin works well with the jvm metrics but  does not parse the nested json maps returned by cassandra metrics properly when the query has wild cards for keyspaces or table names. Since I had to add logic that is specific to Cassandra, I though a cassandra specific plugin will work better.

sparrc commented:
okay, carry on :+1:

subhachandrachandra commented:
Simplified the config to accept both servers and metrics as simple lists of strings.

sparrc commented on plugins/inputs/cassandra/README.md:
````diff
@@ -0,0 +1,30 @@
+
+Given a configuration like:
+
+```ini
````
should be `toml` not `ini`

sparrc commented on plugins/inputs/cassandra/README.md:
````diff
@@ -0,0 +1,30 @@
+Given a configuration like:
+
+```ini
+[inputs.cassandra]
````
this should be `[[inputs.cassandra]]`

sparrc commented on plugins/inputs/cassandra/README.md:
```diff
@@ -0,0 +1,30 @@
+[inputs.cassandra]
+  context = "/jolokia/read"
+  servers = [":8778"]
+  metrics = ["/java.lang:type=Memory/HeapMemoryUsage"]
```
could you provide more examples of useful/common metrics?

sparrc commented on plugins/inputs/cassandra/README.md:
```diff
@@ -0,0 +1,30 @@
+
+#### Description
+
+The Cassandra plugin collects Cassandra/JVM metrics exposed as MBean's attributes through jolokia REST endpoint. All metrics are collected for each server configured.
```
Can you link to any Cassandra documentation on this?

sparrc commented on plugins/inputs/cassandra/cassandra.go:
```diff
@@ -0,0 +1,316 @@
+  ## ReadLatency metrics for all keyspaces and tables.
+  ## "type=Table" in the query works with Cassandra3.0. Older versions might
+  ## need to use "type=ColumnFamily"
+  metrics  = ["/java.lang:type=Memory/HeapMemoryUsage",
```
for line-length, change this to:

```toml
metrics  = [
  "/java.lang:type=Memory/HeapMemoryUsage",
  "/org.apache.cassandra.metrics:type=Table,keyspace=*,scope=*,name=ReadLatency"
]
```

sparrc commented:
thanks @subhachandrachandra, just a few comments to address, and can you also update the README and CHANGELOG files?

subhachandrachandra commented:
Thanks for your comments @sparrc . Addressed all of them and updated README and CHANGELOG. Added a Thank You to myself in the CHANGELOG so you don't have to edit the file again :-)

sparrc commented:
@subhachandrachandra can you rebase your changes off master one more time?

sparrc commented:
@subhachandrachandra your branch and/or fork appears to have diverged a few times from the influxdata master branch. Could you please squash your changes into one commit and rebase them on top of master?

subhachandrachandra commented:
Closing this request. Will create new one after squashing my changes.
